### PR TITLE
Handle empty responses correctly in grpc_http1_reverse_bridge.

### PR DESCRIPTION
**Pull request by hq6**
Previously, when an http 1 server returned a header-only response, the grpc_http1_reverse_bridge bridge would fail to prepend a frame header, because end_stream was ignored in encodeHeader, and encodeData is not called for header-only responses.

That meant grpc clients were missing the expected frame header, resulting in the request failing when it should succeed.

This commit ensures that empty responses have the grpc frame header added.

As part of this change the FilterCallState::LastDataFrame bit is set on the filter_call_state_ when calling encodeHeaders with end_stream=true, because this is the first filter to need to call addDecodedTrailers inside encodeHeaders.

Additionally, `FilterCallState::LastDataFrame` is renamed to `FilterCallState::EndOfStream` to reflect its expanded role.

This commit fixes the following issues:

https://github.com/envoyproxy/envoy/issues/29989

### CODE REVIEW

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I'm not suggesting you should use that return code, but it does document the fact that you cannot currently add trailers to response with no data.

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I guess the part of the documentation that was ambiguous to me is whether the documented restriction applies *only* in the case where `ContinueAndDontEndStream` is returned.

What makes me currently believe that it applies only in this case is the use of the words "in this case":
```
  // TODO(yosrym93): Support adding trailers in this case by implementing new functions to inject
  // trailers, similar to the inject data functions.
```

Thus, I have two questions for you:
1. Are you fairly confident that the restriction applies to *all* cases of header-only response, rather than only the case of a header-only response where the data to be added is not known at the time of encodeHeaders?
2. If so, what negative consequences might result from this change?

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: When I run a debug build of Envoy and an upstream responds with an empty GRPC response, the assertion I linked fails. Release builds disable assertions, so it succeeds.

The negative consequences would be related to the case where there are multiple HTTP filters in use or if multiple filters try to inject trailers. Perhaps @yosrym93 can provide more details?

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: @zuercher I also ran a debug build on the most recent iteration of this PR  and experimented using the Ruby repo linked above. The README on that repo has repro instructions.
 * Without the changes in filter_manager, I also get the assertion failure.
 * With the changes in filter_manager, I no longer get the assertion failure.

**If you ran the most recent version, could you please provide repro instructions? Thanks!**

With respect to negative consequences:
 * As far as I can tell, the restriction on multiple failures trying to inject trailers has always existed and is independent of this change.
 * I'm uncertain about the case with multiple HTTP filters in use.

Comment (yosrym93) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I haven't touched this code in a very long time, so my memory is terrible. I **think** the comment about missing support for injecting trailers is particularly about the case when the trailers will be injected asynchronously (i.e. not during headers iteration).

I am not saying that adding trailers is supported during headers iterations, I can't remember tbh, but that comment was in the context of injecting trailers after headers iteration.

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: Thanks. I'm experimenting with a change that would trigger the body and trailer injection using delayed dispatch outside the encodeHeaders call. I think that might work.

Alternatively, maybe @alyssawilk can say whether it's safe to invoke `addEncodedTrailers` from `encodeHeaders` to set a trailer when we know there's no body. In that case, we'd need to think about what to do with [this assertion](https://github.com/envoyproxy/envoy/blob/main/source/common/http/filter_manager.cc#L1323).

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I think your assertion link might be incorrect, but if you're referring to the assertion inside `addEncodedTrailers`, this PR's changes in `filter_manager.h` already get around that assertion by setting the LastDataFrame bit.

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: Yeah, I just don't know if subverting that assertion makes sense or not.

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I think another integration test (see test/integration/filter_integration_test.cc) is in order. Looks like you could add a parameter to the existing test-only [AddBodyFilter filter](https://github.com/envoyproxy/envoy/blob/main/test/integration/filters/add_body_filter.cc) that controls whether it adds trailers in the header and then a test case to exercise it.

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: (Parameters are defined in the matching add_body_filter.proto, I meant to add.)

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: Also, we need to implement this for the decode path as well, and will need a test there too.

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: I can look into adding the test.

For the decode path, are you asking for a change in this filter or a change somewhere else?
I'm asking because I'm not sure what the use case is in this filter, given that well-formed gRPC requests should never have an empty body.

Comment (zuercher) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: It's not necessary for the grpc filter, but the decode (downstream request)  and encode (upstream response) should be mostly symmetric. Need to allow empty requests to have body and trailers added on the decode path.

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: Is there something special I need to do to get `bazel` to pick up the updated proto?

I added a field in `test/integration/filters/add_body_filter.proto`

```
bool add_trailers_in_encode_decode_header = 4;
```

But getting compiler errors when trying to build:

```
bazel test --sandbox_debug :filter_integration_test
test/integration/filters/add_body_filter.cc: In member function 'virtual Envoy::Http::FilterHeadersStatus Envoy::AddBodyStreamFilter::encodeHeaders(Envoy::Http::ResponseHeaderMap&, bool)':
test/integration/filters/add_body_filter.cc:109:22: error: 'using element_type = class Envoy::AddBodyFilterConfig' {aka 'class Envoy::AddBodyFilterConfig'} has no member named 'add_trailers_in_encode_decode_header_'
  109 |         if (config_->add_trailers_in_encode_decode_header_) {
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```
I tried both the spelling with the trailing `_` and without.

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: @zuercher I have added the requested tests, as well as the symmetric change on the decoder side.

Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: @zuercher Are there additional changes you'd like to see in this PR, or tests that you'd like to have run?

I haven't written an integration test before in Envoy, so if this was not what you had in mind, please do let me know. :)

**Review by zuercher — DISMISSED** [review]
That looks good to me. Thanks for taking the time to do the extra bits beyond what was needed for the reverse bridge filter.

**Comment by zuercher** [comment]
cc @alyssawilk for the filter manager part

**Comment by hq6** [comment]
> That looks good to me. Thanks for taking the time to do the extra bits beyond what was needed for the reverse bridge filter.

@zuercher Thanks for the review and for the thoughtful suggestions!

Just to be explicit, I do not currently have write access to the repository, so I cannot merge this PR myself.

Screenshot 2023-12-04 at 9 15 48 PM

I am hoping that either you or @alyssawilk will merge this PR once she's looked at it, unless there's some procedure I should be following to get write access.

**Comment by zuercher** [comment]
> Just to be explicit, I do not currently have write access to the repository, so I cannot merge this PR myself.

A maintainer will handle merging it. It still needs another review, per our rules.

File: source/common/http/filter_manager.cc
Comment (alyssawilk) [inline source/common/http/filter_manager.cc:582]: hm, I see why you had to do this and it looks like  safe change, but I think you need to update the comment and probably rename for clarity.
/wait

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: is the correct thing to do here to have [headers][trailers] or edit the headers?  cc @lizan

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (alyssawilk) [inline source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:196]: as a funcional change to the daa plane (even a bug fix) this should be runtime guarded and have a release note, thanks!

**Review by alyssawilk — COMMENTED** [review]
thanks for the well tested fix!

File: source/common/http/filter_manager.cc
Comment (hq6) [reply source/common/http/filter_manager.cc:582]: @alyssawilk I'm guessing you're referring to the comment and name of `LastDataFrame`, but if you're not, please let me know.
Proposed change:
```
diff --git a/source/common/http/filter_manager.h b/source/common/http/filter_manager.h
index e859b5bd45..9891cd540e 100644
--- a/source/common/http/filter_manager.h
+++ b/source/common/http/filter_manager.h
@@ -1056,9 +1056,11 @@ private:
       // to verify we do not encode1xx headers more than once per
       // filter.
       static constexpr uint32_t Encode1xxHeaders  = 0x100;
-      // Used to indicate that we're processing the final [En|De]codeData frame,
-      // i.e. end_stream = true
-      static constexpr uint32_t LastDataFrame = 0x200;
+      // Used to indicate that one of the following conditions is true:
+      // 1. The filter manager is processing the final [En|De]codeData frame.
+      // 2. The filter manager is in [En|De]codeHeaders and is processing a
+      //    header-only request or response.
+      static constexpr uint32_t EndOfStream = 0x200;
 
       // Masks for filter call state.
       static constexpr uint32_t IsDecodingMask = DecodeHeaders | DecodeData | DecodeMetadata | DecodeTrailers;
```
Does this seem sufficiently clear, or do you prefer a different name?

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:195]: @alyssawilk For this particular filter, the correct response based on the gRPC protocol is actually [headers][grpc_frame_header][trailers] so I think the trailers are needed.

File: source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc
Comment (hq6) [reply source/extensions/filters/http/grpc_http1_reverse_bridge/filter.cc:196]: @alyssawilk The [runtime guarding documentation](https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md#runtime-guarding) states the following:

> Runtime guarded features may either set true (running the new code by default) in the initial PR, after a testing interval, or during the next release cycle, at the PR author's and reviewing maintainer's discretion.

Do you have a preference for whether this is set to true by default?

**Comment by hq6** [comment]
@alyssawilk I believe I have addressed all your comments, but please let me know if I've missed anything.

**Review by alyssawilk — APPROVED** [review]

**Review by zuercher — APPROVED** [review]

**Pull request merged**
